### PR TITLE
Faking performance when not present should throw an error (and minor changes) - fix for #374 

### PR DESCRIPTION
**Pull request by itayperry**
#### Original purpose: enabling adding performance to toFake array - apparently this was already possible so a few minor changes were made.

Issue #374 refers to the fact that `fake-timers` overrides `performance` object (as in other properties/methods in addition to overriding `performance.now()`).  The first suggested solution/notion was to enable users to add `performance` to the `toFake` array - after realizing this was already possible a test was added to make sure an error is thrown when a user puts `performance` in the `toFake` array but the `performance` object isn't present.

I don't know if this PR makes a good addition/improvement - I'm waiting for your notes and comments and I'll fix whatever's needed.

### CODE REVIEW

File: src/fake-timers-src.js [hunk 1]
```diff
@@ -1613,6 +1599,25 @@ function withGlobal(_global) {
             clock.attachedInterval = intervalId;
         }
 
+        if (clock.methods.includes("performance")) {
```
[the remark this answers is not in the document]
Comment (itayperry) [reply src/fake-timers-src.js:1602]: I tested it now and it's not exactly the same - many tests are failing, I think it's due to this piece of code that runs earlier:

```js
        clock.methods = config.toFake || [];

        if (clock.methods.length === 0) {
            // do not fake nextTick by default - GitHub#126
            clock.methods = Object.keys(timers).filter(function (key) {
                return key !== "nextTick" && key !== "queueMicrotask";
            });
        }
```

if `config.toFake` is empty I still need to stub `performance` in many tests cause the `clock.methods` array will later have it

> ...and at the same time flatten the conditionals by throwing early

I still wanna do that for the code to be cleaner - maybe I'll separate the conditions entirely? But the code still ain't pretty I'm afraid.

As in, doing the first one that throws in here:

```js
clock.methods = config.toFake || [];

if (
    config.toFake &&
    config.toFake.includes("performance") &&
    !hasPerformancePrototype
) {
    // user explicitly tried to fake performance when not present
    throw new ReferenceError(
        "non-existent performance object cannot be faked"
    );
}

if (clock.methods.length === 0) {
    // do not fake nextTick by default - GitHub#126
    clock.methods = Object.keys(timers).filter(function (key) {
        return key !== "nextTick" && key !== "queueMicrotask";
    });
}
```
and later do this:

```js
        if (clock.methods.includes("performance") && hasPerformancePrototype) {
            var proto = _global.Performance.prototype;
            Object.getOwnPropertyNames(proto).forEach(function (name) {
                if (name !== "now") {
                    clock.performance[name] =
                        name.indexOf("getEntries") === 0 ? NOOP_ARRAY : NOOP;
                }
            });
        }
```

[the remark this answers is not in the document]
Comment (fatso83) [reply src/fake-timers-src.js:1602]: AFAIK, the problem you have is that the default methods to fake includes stuff that is not necessarily supported, right? If so, why not just create a general `isSupported` map that you can check in that loop? Something like :

```javascript
const isSupported = new Map();
isSupported.set("performance", hasPerformancePrototype);
isSupported.set("nextTick", hasNextTick);
...
const supported = key => !isSupported.has(key) || isSupported.get(key) // true for all not explicitly added
if (clock.methods.length === 0) {
  clock.methods = Object.keys(timers).filter(supported).filter(function (key) {
      return key !== "nextTick" && key !== "queueMicrotask";
  });
}
```
That should filter out performance and other unsupported stuff from the default list of methods to fake if `toFake` was empty.

[the remark this answers is not in the document]
Comment (itayperry) [reply src/fake-timers-src.js:1602]: I have another question - regarding the suggested code above:

```js
if (config.toFake.includes("performance") {
            if (!hasPerformancePrototype) { 
                    throw new ReferenceError("non-existent performance object cannot be faked");
            }
            // code to stub out performance follows 
            var proto = _global.Performance.prototype;        
            ...
}
```

It ignores the situation in which `toFake` is empty and then the entire `var proto = _global.Performance.prototype;` does not run, and it should in many cases - the `isSupported` `Map`  does not seem to deal directly with this problem, although it's a really cool and neat idea, and maybe I should do it anyway.

File: test/fake-timers-test.js [hunk 2]
```diff
@@ -3700,6 +3700,21 @@ describe("FakeTimers", function () {
             });
         }
 
+        if (!performanceNowPresent) {
```
[the remark this answers is not in the document]
Comment (itayperry) [reply test/fake-timers-test.js:3703]: > use Mocha's built-ins to skip tests

You're so right! I've used this before and I should have known better than to forget about it :)

> We have the possibility of choosing what the target is

I didn't think about it, it's really cool! I can't seem to get it to work exactly like that but I cheated a little lol and stole properties from another test that uses a `global`/`target` and it finally worked!

```js
it("throws when adding performance to tofake array when performance not present", function () {
    try {
        const setTimeoutFake = sinon.fake();
        const context = {
            Date: Date,
            setTimeout: setTimeoutFake,
            clearTimeout: sinon.fake(),
            performance: undefined
        };
        FakeTimers.withGlobal(context).install({
            toFake: ["performance"],
        });
    } catch (e) {
        const expectedMsg =
            "non-existent performance object cannot be faked";
        assert(e instanceof ReferenceError);
        assert.equals(e.message, expectedMsg);
    }
});
```

Is this ok?

[the remark this answers is not in the document]
Comment (fatso83) [reply test/fake-timers-test.js:3703]: Hi, @itayperry . Sorry, for leaving you hanging for more than a month! I was not aware I was blocking it. This seems good to me!

With regards to the assertion, I would check out the exception assertion in referee: https://sinonjs.github.io/referee/#exception

[the remark this answers is not in the document]
Comment (itayperry) [reply test/fake-timers-test.js:3703]: I was too insecure to send a message lol I'm on it. I'll do it as fast as I can :)

> I would check out the exception assertion in referee 

Looks like best practice! Thank you :)

**Comment by itayperry** [comment]
Hi @fatso83 :) I didn't edit the README file, but I still get a `prettier` error because of it - should I fix that and just make another commit?

>  [warn] README.md

**Comment by fatso83** [comment]
Yup, that was me editing the README in the web UI.

**Review by fatso83 — APPROVED** [review]
Finally mergeable! And in time for X-Mas! 😸 Great stuff, Itay.

**Comment by itayperry** [comment]
For glory, for fake-timers! lol :)

**Pull request merged**